components/learn/modules/ROOT/pages/on-upgrades.adoc

```diff
@@ -1,8 +1,281 @@
+
+contract AdminBox is Initializable {
+    uint256 private value;
+    address private admin;
+
+    function initialize(address _admin) public initializer {
+        admin = _admin;
+    }
+
+    // Stores a new value in the contract
+    function store(uint256 newValue) public {
+        require(msg.sender == admin);
+        value = newValue;
+        emit ValueChanged(newValue);
+    }
+
+    // Reads the last stored value
+    function retrieve() public view returns (uint256) {
+        return value;
+    }
+}
+```
+
+When deploying this contract, the CLI will prompt us to execute the initializer and ask us to provide the admin address. 
+
+```bash
+$ npx oz create
+✓ Compiled contracts with solc 0.5.9
+? Pick a contract to instantiate: AdminBox
+? Pick a network: development
+✓ Contract AdminBox deployed
+? Call a function to initialize the instance after creating it? Yes
+? Select which function: initialize(_admin: address)
+? _admin (address): 0x90f8bf6a479f320ead074411a4b0e7944ea8c9c1
+✓ Setting everything up to create contract instances
+✓ Instance created at 0x2612Af3A521c2df9EAF28422Ca335b04AdF3ac66
+```
+
+For all practical purposes, the initializer acts as a constructor. However, keep in mind that, since it's a regular function, you need to manually call the initializers of all base contracts, if any.
+
+NOTE: In future versions of the OpenZeppelin SDK, the CLI will take care of automatically converting constructors into initializers, so you do not need to worry about this.
+
+To learn more about this and other caveats when writing upgradeable contracts, check out our xref:sdk::writing-contracts.adoc[Writing Upgradeable Contracts] guide.
+
+=== Upgrading
+
+Due to technical limitations, when you upgrade a contract to a new version, you cannot change the **storage layout** of that contract.
+
+This means that, if you have already declared a state variable in your contract, you cannot remove it, change its type, or declare another variable before it. In our `Box` example, it means that we can only add new state variables _after_ `value`.
+
+```solidity
+// contracts/Box.sol
+contract Box {
+    uint256 private value;
+
+    // We can safely add a new variable after the ones we had declared
+    address private owner; 
+    
+    // ...
+}
+```
+
+Fortunately, this limitation only affects state variables. You can change the contract's functions and events as you wish.
+
+NOTE: If you accidentally mess up with your contract's storage layout, the CLI will warn you when you try to upgrade.
+
+To learn more about this limitation, head over to the xref:sdk::writing-contracts.adoc#modifying-your-contracts[Modifying Your Contracts] guide.
+
+[[upgrading-contracts-in-js]]
+== Upgrading Contracts Programmatically
+
+If you want to create and upgrade contracts from your JavaScript code instead of via the command line, you can use the `@openzeppelin/upgrades.js` library instead of the CLI.
```

Comment:
```suggestion
If you want to create and upgrade contracts from your JavaScript code instead of via the command line, you can use the *OpenZeppelin Upgrades* library instead of the CLI.
```